standalone-metastore/metastore-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java

```diff
@@ -0,0 +1,139 @@
+import org.apache.hadoop.hive.metastore.SecureServletCaller;
+import org.apache.hadoop.hive.metastore.ServletSecurity;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.apache.iceberg.HiveCachingCatalog;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.hive.HiveCatalog;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.handler.gzip.GzipHandler;
+import org.eclipse.jetty.servlet.ServletContextHandler;
+import org.eclipse.jetty.servlet.ServletHolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class HMSCatalogServer {
+  private static final String CACHE_EXPIRY = "hive.metastore.catalog.cache.expiry";
+  /**
+   * The metric names prefix.
+   */
+  static final String HMS_METRIC_PREFIX = "hmscatalog.";
+  private static final Logger LOG = LoggerFactory.getLogger(HMSCatalogServer.class);
+  private static Reference<Catalog> catalogRef;
+
+  static Catalog getLastCatalog() {
+    return catalogRef != null ? catalogRef.get() :  null;
+  }
+
+  private HMSCatalogServer() {
+    // nothing
+  }
+
+  public static HttpServlet createServlet(SecureServletCaller security, Catalog catalog) throws IOException {
+    try (HMSCatalogAdapter adapter = new HMSCatalogAdapter(catalog)) {
+      return new HMSCatalogServlet(security, adapter);
+    }
+  }
+
+  public static Catalog createCatalog(Configuration configuration) {
+    final String curi = configuration.get(MetastoreConf.ConfVars.THRIFT_URIS.getVarname());
+    final String cwarehouse = configuration.get(MetastoreConf.ConfVars.WAREHOUSE.getVarname());
+    final String cextwarehouse = configuration.get(MetastoreConf.ConfVars.WAREHOUSE_EXTERNAL.getVarname());
+    final HiveCatalog catalog = new org.apache.iceberg.hive.HiveCatalog();
+    catalog.setConf(configuration);
+    Map<String, String> properties = new TreeMap<>();
+    if (curi != null) {
+      properties.put("uri", curi);
+    }
+    if (cwarehouse != null) {
+      properties.put("warehouse", cwarehouse);
+    }
+    if (cextwarehouse != null) {
+      properties.put("external-warehouse", cextwarehouse);
+    }
+    catalog.initialize("hive", properties);
+    long expiry = configuration.getLong(CACHE_EXPIRY, 60_000L);
+    return expiry > 0? HiveCachingCatalog.wrap(catalog, expiry) : catalog;
+  }
+
+  public static HttpServlet createServlet(Configuration configuration, Catalog catalog) throws IOException {
+    String auth = MetastoreConf.getVar(configuration, MetastoreConf.ConfVars.CATALOG_SERVLET_AUTH);
+    boolean jwt = "jwt".equalsIgnoreCase(auth);
+    SecureServletCaller security = new ServletSecurity(configuration, jwt);
+    Catalog actualCatalog = catalog;
+    if (actualCatalog == null) {
+      actualCatalog = createCatalog(configuration);
+      actualCatalog.initialize("hive", Collections.emptyMap());
+    }
+    catalogRef = new SoftReference<>(actualCatalog);
+    return createServlet(security, actualCatalog);
```

Comment:
should we put the `catalogRef` to `createServlet` instead of `actualCatalog`, so no strong reference to the `actualCatalog`?

Reply:
No, the servlet needs to hold a strong ref; the server holds it as a static soft ref for testing convenience.